Variables: Explicitly type variable editor extended state

joshhunt opened the pull request:
**What this PR does / why we need it**:

Supersedes https://github.com/grafana/grafana/pull/44687.

Updates the `template.editor` state type to explicitly type the `extended` properties to be a union to avoid the type assertions in the views.

Adds new selectors to narrow these extended properties down to the specific type for that component, rather than using `as` type assertions.

_Later_ ([in a seperate PR](https://github.com/grafana/grafana/pull/44957)) we'll refactor the actions to type them and dispatch atomic changes (rather than one property at a time) to fix an out-of-sync bug

 - [x] Update `template.editor` state types
 - [x] Update views
 - [x] Update/add tests

**Which issue(s) this PR fixes**:

Relates to #43803

hugohaggmark commented on public/app/features/variables/adhoc/AdHocVariableEditor.tsx:
```diff
@@ -38,8 +36,8 @@ export class AdHocVariableEditorUnConnected extends PureComponent<Props> {
     const { variable, editor } = this.props;
-    const dataSources = editor.extended?.dataSources ?? [];
-    const infoText = editor.extended?.infoText ?? null;
+    const dataSources = (editor.extended?.kind === 'adhoc' && editor.extended?.dataSources) || [];
```
Hmm, where does kind come from, the variable already has a type so I don't think we need another kind/type. And I think we should create guards for this instead in `public/app/features/variables/guard.ts`.

joshhunt replied:
`kind` is the important discriminator to distinguish between the different types of extended. We need some way to (statically) determine which kind of ExtendedState it is.

Importantly, `kind` is not a string, but is a union of string constants, so we get pretty strong type safety around them.

```typescript
typeof<kind> === "adhoc" | "query" | "datasource"
```

Also, it's [not possible to access incorrect properties](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIIBMASB7BA1OKYOAIwBsIBRdYMbKAZTDkmQG8AoZbga1HQBcyAORx0AC1zCA3F26gY2ACoQAHmCEBnMERABzWd2ToWcBtgCuUJJqGooUOAE8APG2SR1WnaAPIAbnBkFhDeunrIAD7IIBZkZMgAvgB8hokcXKCQsIgoACKm5lZIBESkFNS09EwsKJxGfCCCIibMmpbWEDJyxoUdSEpOAA4QtmgOzm4eahrI2uHSAUEhYb5JqVzpmeDQ8EjIAIohUE6lxORUNHSMzKz1vPxCwgCOx07dRmflEEfQTpXXIQWEA8EDYADuIEM3FaZn6oTmPn0URicTIaQyHmGKC+FwB9Eo6ggIE0wGwIGQAF40FhcLiKldqrcUNECswip16ZcqjdaijficufjeZBZBwYMCEGAyRSoMT0NAABQzeUQZpCxlQQmQEkygCU7ExRgA9MbkCqmmrkMBNOA5iMEMAYMAEOawRY9OIADRzbDIBBwEBgsDIRA2TTIIZQbAjKDS0bIcnWsA9f3k9oUAB0ZGwemVRMt6EzCmUMz1YqM1pgyHzOoVRca6CplOpACJXn9Wwb7pXuKbzQX68hxHAI2DwcgSBBiTFCNHwVa6Ic3hqeTVID72sgF-7A8HQwhwx5xCgMJJXVGY9B45pU5WEOnsFmc3mLfXiyBFCp1OW70YHyST4QNmua1qqRawhyNi-pW6SJEAA) for the type.

I understand what you say about the variable already having a type, but due to the structure of the state (that type isn't in the Editor state) it can't be used to type these extended properties. The state could (theoretically) be rearranged to put everything together and we could make a single interface for each type of variable, but that all seems like a lot more work.

Another option could be use [`property in object` style narrowing](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIIBMASB7BA1OKYOAIwBsIBRdYMbKAZTDkmQG8AoZb0GbAFQgAPMAC5kAZzBEQAcwDcXbuhZwG2AK5QkE8aihQ4ATwA8bZJBHipMhcgBucMhojXpoWcgA+yEBrJkyAC+AHyKwRxcoJCwiCgAIqrqWkgERKQU1LT0TCwonNzIKszJ2hD8RgAOELpoBsZmFsJiku5y8g5OLm62wWFcQZHI0dDwSMgAii5QRmnE5FQ0dIzMrAXccxkQU9BGWcviGiAA1iDYAO4g4cpJmmU9Ht6+-mThg1xgVSibC-v0lCIICAJMBsCBkABeNBYXA-TJLHKrFA+RIlO6pQjzeHZFZ5J47GZwxY43KQRQcGBHBBgUHgqBA9DQAAUzQZEHQ4iJfygAMgwNpAEp2ENCgB6UXIVkgRnoYYScCSaoIYAwYAISVnDSyAAWABpJNhkAg4CAzmBkIgdBJkJUoNhqlAaTVkGDhuajBAwEpuAgwRJsBQAHRkbCyFmA6XswO8ATNAXkwrDGDIJkAImKanRNVTw3BUplQvWie44slEZlyG1cGtZ3OyBIECBvkIdvO7IshowOHwmK23NJEG9xd9wIDEGDofDfJl0ZAfEEInjQ8TI-9QZDYfzUYzpR0S8TgyCQA) using properties that are unique to each subtype, but I prefer the method here - I think the explicitness makes it easier to understand what's going on.

joshhunt replied:
I'm also wanted to try and avoid user-defined type guards (`function isAdhoc(obj): obj is AdhocExtendedState`) as they're just another kind of "as" type cast/assertion.

hugohaggmark replied:
I don't agree with you, we have type safety today when we consume the state, but not when we dispatch actions. I don't agree with that the introduction of the `kind` will hinder you from dispatching the wrong editor action. I think the property in object style narrowing will suffice for this.

hugohaggmark reviewed (commented):
Not sure about the new `kind` prop. Variables already have `type` so we should be able to deduce the VariableEditorState type from the variable type with a guard instead.

hugohaggmark commented on public/app/features/variables/adhoc/AdHocVariableEditor.tsx:
```diff
@@ -5,15 +5,13 @@ import { DataSourceRef, SelectableValue } from '@grafana/data';
 import { VariableSelectField } from '../editor/VariableSelectField';
 
 const mapStateToProps = (state: StoreState) => ({
-  editor: state.templating.editor as VariableEditorState<AdHocVariableEditorState>,
```
Could we create a guard like function that takes in state and variable and that returns the correct state type here instead.

joshhunt replied:
Yeah, I wanted to take a look at making that easier. I was thinking of something like:

```typescript
function narrowToAdhoc(editorState: VariableEditorState): VariableEditorState<AdhocVariableEditorState> | null {
  return editorState.extended.kind === "adhoc" ? editorState : null; 
}
```

Related: how do we handle the edge case of the AdhocVariableEditor is passed a non-adhoc editor state (somehow??). At the moment it's just continuing on and setting default values, but I'm wondering if it should just "give up" and render null instead? Not sure.

hugohaggmark replied:
If that happens we have problems in other parts of the variable system, so I don't think you need to look into that in this PR.

joshhunt replied:
I guess the key part is that if we have problems in our code for the variable system we want to know that before we even run Grafana. That's the whole point of using Typescript. This is just handling one of those edge/error cases.

hugohaggmark replied:
I'm very aware of why we use TypeScript but I think solving a polymorphic architecture with Redux will always be a challenge that comes with some compromises, this is one of them.

joshhunt commented on public/app/features/variables/editor/selectors.ts:
```diff
@@ -0,0 +1,37 @@
+  VariableEditorState,
+} from './reducer';
+
+export function getAdhocVariableState(
+  editorState: VariableEditorState
+): VariableEditorState<AdHocVariableEditorState> | null {
+  const { extended, ...rest } = editorState;
+
+  if (extended && 'dataSources' in extended) {
+    return {
+      extended,
+      ...rest,
+    };
+  }
+
+  return null;
+}
```
Produces a _whole_ state object that matches a particular subtype. Ideally for use in a `mapStateToProps` function.

joshhunt replied:
(note I did [change this function](https://github.com/grafana/grafana/pull/44749/commits/eab9ecfb8ac6ddce46067227c4077dc9521db40a) slightly in a commit after this to return `VariableEditorState<null>` instead of just `null` when the type does not match)

joshhunt commented on public/app/features/variables/editor/selectors.ts:
```diff
@@ -0,0 +1,37 @@
+  return null;
+}
+
+export function getDatasourceVariableState(editorState: VariableEditorState): DataSourceVariableEditorState | null {
+  if (editorState.extended && 'dataSourceTypes' in editorState.extended) {
+    return editorState.extended;
+  }
+
+  return null;
+}
```
Or, a "simpler" selector, more geared to be used in a render function.

hugohaggmark replied:
I prefer this selector over the one used in Adhoc.

joshhunt replied:
How do you feel about then how it's actually used in the components? https://github.com/grafana/grafana/pull/44749/files#diff-7c68d8e773d0949597127594a57f6dc632f7194755786fe546fd37f52f6a9eeb

 - AdhocVariableEditor uses a more complicated selector, but allows the [view _using_ it](https://github.com/grafana/grafana/pull/44749/files#diff-79c3921c3bfa84fefe087b545132b25ff88852dbd8b26dd26efb8e6f3743156b) to be simpler and less room for mistakes
   - however this one require a [bunch of generics in the state type](https://github.com/grafana/grafana/pull/44749/files#diff-4a995a538d05b0805440eca25f743a6521027148fec55afcd1ba60e57a97c40aR22-R24) which makes it more complicated
 - DatasourceVariableEditor's selector is simpler, but [requires the guard in more places in the view](https://github.com/grafana/grafana/pull/44749/files#diff-7c68d8e773d0949597127594a57f6dc632f7194755786fe546fd37f52f6a9eeb) and/or more changes to the component
   - could be mitigated by using it in the `mapStateToProps` function with a new mapped prop


thoughts?

joshhunt reviewed (commented):
I've updated this partially to show two other different takes on utilities producing a narrowed variable editor state object using typescript's natural type narrowing, rather than type holes like user-defined type guards or type assertions, while removing the `kind` discriminator from before.

 - `AdHocVariableEditor.tsx` uses a selector function in `mapStateToProps` to narrow `VariableEditorState<VariableEditorExtension>` to `VariableEditorState<AdHocVariableEditorState>`. This keeps the render function basically identical to how it is currently in main.
 - `DatasourceVariableEditor.tsx` uses a utility function in render() to return just specific datasource variable extended props the generic editor state.

I think I prefer the first approach (in AdHocVariableEditor) more as it keeps most of the component basically the same as before - it just replaces the previous type assertion with a naturally narrowing selector.

hugohaggmark commented on public/app/features/variables/query/QueryVariableEditor.tsx:
```diff
@@ -115,7 +113,13 @@ export class QueryVariableEditorUnConnected extends PureComponent<Props, State>
+
+    if (
+      !editor.extended ||
+      editor.extended.kind !== 'query' ||
```
There is no kind right? Why does this work?

joshhunt replied:
It almost definitely doesn't work 😅. Just did new style in a few places to demo it.

this PR is all _very_ incomplete, still draft work-in-progress

hugohaggmark reviewed (commented):
Great work on this! I think this is a great step forward. I left some comments and I'm still missing tests before I can approve it 😉

hugohaggmark commented on public/app/features/variables/datasource/DataSourceVariableEditor.tsx:
```diff
@@ -108,19 +115,21 @@ export class DataSourceVariableEditorUnConnected extends PureComponent<Props> {
-const mapStateToProps: MapStateToProps<ConnectedProps, OwnProps, StoreState> = (state, ownProps) => ({
-  editor: state.templating.editor as VariableEditorState<DataSourceVariableEditorState>,
-});
+const mapStateToProps: MapStateToProps<ConnectedProps, OwnProps, StoreState> = (state, ownProps) => {
+  return {
+    editor: state.templating.editor,
+  };
+};
```
While we're in the mood of changing these things we should make sure we follow the new guide lines here:https://github.com/grafana/grafana/blob/main/contribute/style-guides/redux.md#typing-of-connected-props

joshhunt committed:
Narrow Variable editor state using selector functions

 - Explicitly type "extended" editor state in editor/reducter.ts using a union
 - Create selectors to narrow the types, using unique properties from each extended state to discriminate the union
 - Update DataSourceVariableEditor to use new style of redux connector
 - Update variable editor components to use new selectors

joshhunt committed:
fix tests

joshhunt committed:
Make adhoc variable infoText optional, because it is!

joshhunt committed:
Add AdHocVariableEditor tests

joshhunt committed:
DataSourceVariableEditor tests

joshhunt committed:
comment

joshhunt committed:
reset

joshhunt committed:
Wrote tests for selectors \(that actually caught a bug, whodathunkit)

joshhunt committed:
fix stray types and lint issues

hugohaggmark commented on public/app/features/variables/editor/selectors.ts:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Narrows generic variable editor state down to specific Adhoc variable extended editor state
+ */
+export function getAdhocVariableState(editorState: VariableEditorState): AdHocVariableEditorState | null {
```
super small nit: can we name these selectors `get<Type>EditorState` instead?

joshhunt replied:
Totally - I haven't thought too much about naming in this at all, and was hoping you would pick some things up!

joshhunt committed:
Rename selector functions

hugohaggmark commented on public/app/features/variables/editor/reducer.test.ts:
```diff
@@ -188,6 +188,7 @@ describe('variableEditorReducer', () => {
         .thenStateShouldEqual({
           ...initialVariableEditorState,
           extended: {
+            // @ts-ignore - temp ignoring this, we'll fix it soon
```
Hmm, this looks weird

joshhunt replied:
Because `extended` is now explicitly typed more than just `{}`, `someProp` is no longer a valid here. In hindsight I'm not sure why I didn't just fix this test in this PR, but I did [fix it as intended in the next PR](https://github.com/grafana/grafana/pull/44957/files#diff-c48541823dee41f59d8b2790e7ddb908eb8455755f7b379c53e96ae277a4f41eL191).

I'm hoping we could just leave this as is "for now" given I fix it immediately in the next PR?

hugohaggmark replied:
Thanks for the explanation and I've approved the PR anyway 👍

hugohaggmark commented on public/app/features/variables/editor/reducer.ts:
```diff
@@ -65,6 +82,7 @@ const variableEditorReducerSlice = createSlice({
       state: VariableEditorState,
       action: PayloadAction<{ propName: string; propValue: any }>
     ) => {
+      // @ts-ignore - temp ignoring the errors now the state type is more strict
```
Hmm, is this necessary?

joshhunt replied:
It is, because `extended` cannot be an empty object (with superfluous properties).

`state.extended` is nullable, and spreading null into an object gives you an empty object. Typescript then complains that we aren't setting `extended` to `AdHocVariableEditorState | DataSourceVariableEditorState | QueryVariableEditorState`, which we aren't because `propName` is a string (so we/ts don't know what it is).

This is how this state is currently untyped, so I don't think this regresses anything. It just ignores an error we weren't (but should have been) catching before.

Fixing this is subject of the [whole separate PR](https://github.com/grafana/grafana/pull/44957/files#diff-4a995a538d05b0805440eca25f743a6521027148fec55afcd1ba60e57a97c40aL85) which should all be ready to go in at the same time, so I'm hoping this will fly just for a few minutes 😇

hugohaggmark replied:
Thanks for the explanation and I've approved the PR anyway 👍

hugohaggmark reviewed (approved):
Great work on this 🎉  Found some `@ts-ignore` that seemed weird at first glance but maybe they are necessary now?
